internal/archive/credentials.go

```diff
@@ -0,0 +1,304 @@
+package archive
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// This file defines findCredentials() function for searching repository
+// credentials in Apt configuration, see
+// https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html.
+
+// credentials contains matched non-empty Username and Password.
+// Username is left empty if the search is unsuccessful.
+type credentials struct {
+	Username string
+	Password string
+}
+
+// Empty checks whether c represents unsuccessful search.
+func (c credentials) Empty() bool {
+	return c.Username == ""
+}
+
+// credentialsQuery contains parsed input URL data used for search.
+type credentialsQuery struct {
+	scheme     string
+	host       string
+	port       string
+	path       string
+	needScheme bool
+}
+
+// queryFor parses repoUrl into credentialsQuery and fills provided credentials with
+// username and password if they are specified in repoUrl.
+func queryFor(repoUrl string, creds *credentials) (*credentialsQuery, error) {
```

Comment:
I get your suggestion, but doing it at once allow us to 1) parse the URL only once and 2) skip allocating `credentialsQuery` if there are credentials in the URL. The function was merged in https://github.com/canonical/chisel/pull/60 as 
```
func parseRepoURL(repoURL string) (creds *credentials, query *credentialsQuery, err error) {
```
which is somewhat better that what it used to be, though it still feels to be doing more than the name suggests.

Since it is already merged as such, I'll resolve this. If you feel it should be modified anyway, feel free to re-open.